File: 2020-week1/c++/2020-week1_18.cpp
#include <iostream>
#include <string>

using namespace std;

class Tree {
	string data;
	Tree *left, *right;
	
public:
	Tree() {
		data= "";
		left = NULL;
		right = NULL;
	}
	
	void setData(char data) {
		this->data = data;
	}
	void setLeft(Tree *left) {
		this->left = left;
	}
	void setRight(Tree *right) {
		this->right = right;
	}
	
	// 전위 
	void static preorder(Tree *root){
		if(root){
			cout << root->data; 
			preorder(root->left);
			preorder(root->right);
		}
	}
	
	// 중위 
	void static inorder(Tree *root){
		if(root){
			inorder(root->left);
			cout << root->data;
			inorder(root->right);	
		}
	}
	
	// 후위 
	void static postorder(Tree *root){
		if(root){
			postorder(root->left);
			postorder(root->right);
			cout << root->data;
		}
	}
};


int main() {
	int n;
	cin >> n;
	
	Tree *tree = new Tree[n];
	
	for(int i=0; i<n; i++){
		char data, left, right;
		cin >> data >> left >> right;
		
		if(data != '.'){
			tree[(int)(data-'A')].setData(data);
		}
		
		if(left != '.'){
			tree[(int)(data-'A')].setLeft(&tree[(int)(left-'A')]);
		}
		else{
			tree[(int)(data-'A')].setLeft(NULL);
		}
		
		if(right != '.'){
			tree[(int)(data-'A')].setRight(&tree[(int)(right-'A')]);
		}
		else{
			tree[(int)(data-'A')].setRight(NULL);
		}
	}
	
	Tree *root = &tree[0];
	Tree::preorder(root);
	printf("\n");
	Tree::inorder(root);
	printf("\n");
	Tree::postorder(root);
	printf("\n");
	 
	return 0;
}

File: 2020-week1/java/2020-week1_18.java
import java.util.Scanner;

public class TreeTraversal_1991 {
	public static int findParentIndex(String[] tree, String parent) {
		int index = 0;
		
		for(int i=1; i<tree.length; i++) {
			if(tree[i] == null) continue;
			if(tree[i].equals(parent)) {
				index = i;
				break;
			}
		}
		return index;
	}
	public static void preorder(String[] tree, int index) {
		if(tree[index] == null) return ;
		
		System.out.print(tree[index]);
		preorder(tree, index*2);
		preorder(tree, index*2+1);
	}
	public static void inorder(String[] tree, int index) {
		if(tree[index] == null) return ;
		
		inorder(tree, index*2);
		System.out.print(tree[index]);
		inorder(tree, index*2+1);
	}
	public static void postorder(String[] tree, int index) {
		if(tree[index] == null) return ;
		
		postorder(tree, index*2);
		postorder(tree, index*2+1);
		System.out.print(tree[index]);
	}
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		
		String[] tree = new String[1000];
		tree[1] = "A";
		
		for(int i=0; i<n; i++) {
			String parent = sc.next();
			String leftChild = sc.next();
			String rightChild = sc.next();
			
			int parentIndex = findParentIndex(tree, parent);
			if(!leftChild.equals(".")) tree[parentIndex*2] = leftChild;
			if(!rightChild.equals(".")) tree[parentIndex*2+1] = rightChild;		
		}
		

		preorder(tree, 1);
		System.out.println();
		inorder(tree, 1);
		System.out.println();
		postorder(tree, 1);	
	}
}
